homeassistant/components/homematicip_cloud/sensor.py

```diff
@@ -183,10 +187,74 @@ async def async_setup_entry(
                     if ch.currentPowerConsumption is not None:
                         entities.append(HmipEsiLedCurrentPowerConsumption(hap, device))
                     entities.append(HmipEsiLedEnergyCounterHighTariff(hap, device))
+        if isinstance(device, AsyncFloorTerminalBlock12):
+            for channel in device.functionalChannels:
+                if isinstance(channel, FloorTerminalBlockMechanicChannel):
+                    if getattr(channel, "valvePosition", None) is not None:
+                        entities.append(
+                            HomematicipFloorTerminalBlockMechanicChannelValve(
+                                hap, device, channel=channel.index
+                            )
+                        )
 
     async_add_entities(entities)
 
 
+class HomematicipFloorTerminalBlockMechanicChannelValve(
+    HomematicipGenericEntity, SensorEntity
+):
+    """Representation of the HomematicIP floor terminal block."""
+
+    _attr_native_unit_of_measurement = PERCENTAGE
+    _attr_state_class = SensorStateClass.MEASUREMENT
+
+    def __init__(
+        self, hap: HomematicipHAP, device, channel, is_multi_channel=True
+    ) -> None:
+        """Initialize floor terminal block 12 device."""
+        super().__init__(
+            hap,
+            device,
+            channel=channel,
+            is_multi_channel=is_multi_channel,
+            post="Valve Position",
+        )
+
+    @property
+    def icon(self) -> str | None:
+        """Return the icon."""
+        if super().icon:
+            return super().icon
+        if hasattr(
```

Comment:
hasattr check can be removed, because the attribute is always available

Reply:
If I removed it, I get this error:

> FAILED tests/components/homematicip_cloud/test_sensor.py::test_hmip_floor_terminal_block_mechanic_channel_1_valve_position - AttributeError: 'HeatDemandChannel' object has no attribute 'valveState'

Reply:
I was really confused why that error happened. The HeadDemandChannel could not be a HomematicipFloorTerminalBlockMechanicChannelValve Instance, because we defined in the async_setup_entry which channels are should be added.

Currently at a few points in the code (not just yours) the functionalChannels Array is accessed via the index (functionalChannels[channelIndex] which gets not necessarily the right functional channel, if there are more then 10 channels. 

The best way until I fixed it is 
```
channel = next(channel for channel in self._device.functionalChannels
                    if channel.index == self._channel)
```

It's untested, but you can use that everywhere, where you are calling `self._device.functionalChannels[self._channel]` at the moment.

Reply:
So like this? https://github.com/home-assistant/core/pull/122759#pullrequestreview-2370369548